stdlib/universe/sort2.go

```diff
@@ -0,0 +1,370 @@
+package universe
+
+import (
+	"container/heap"
+	"context"
+	"sort"
+
+	"github.com/apache/arrow/go/arrow/array"
+	"github.com/apache/arrow/go/arrow/memory"
+	"github.com/influxdata/flux"
+	"github.com/influxdata/flux/arrow"
+	"github.com/influxdata/flux/execute"
+	"github.com/influxdata/flux/internal/arrowutil"
+	"github.com/influxdata/flux/internal/execute/table"
+	"github.com/influxdata/flux/internal/mutable"
+	"github.com/influxdata/flux/plan"
+)
+
+type sortTransformation2 struct {
+	execute.ExecutionNode
+	d       *execute.PassthroughDataset
+	mem     memory.Allocator
+	cols    []string
+	compare arrowutil.CompareFunc
+}
+
+func newSortTransformation2(id execute.DatasetID, spec *SortProcedureSpec, a execute.Administration) (execute.Transformation, execute.Dataset, error) {
+	t := &sortTransformation2{
+		d:       execute.NewPassthroughDataset(id),
+		mem:     a.Allocator(),
+		cols:    spec.Columns,
+		compare: arrowutil.Compare,
+	}
+	if spec.Desc {
+		// If descending, use the descending comparison.
+		t.compare = arrowutil.CompareDesc
+	}
+	return t, t.d, nil
+}
+
+func (s *sortTransformation2) Process(id execute.DatasetID, tbl flux.Table) error {
+	sortCols := make([]int, 0, len(s.cols))
+	for _, col := range s.cols {
+		if idx := execute.ColIdx(col, tbl.Cols()); idx >= 0 {
+			// If the sort key is part of the group key, skip it anyway.
+			// They are all sorted anyway.
+			if tbl.Key().HasCol(col) {
+				continue
+			}
+			sortCols = append(sortCols, idx)
+		}
+	}
+
+	mh := &sortTableMergeHeap{
+		cols:     tbl.Cols(),
+		key:      tbl.Key(),
+		sortCols: sortCols,
+		compare:  s.compare,
+	}
+	if err := tbl.Do(func(cr flux.ColReader) error {
+		return s.processView(mh, cr)
+	}); err != nil {
+		return err
+	}
+
+	out, err := mh.Table(s.mem)
+	if err != nil {
+		return err
+	}
+	return s.d.Process(out)
+}
+
+func (s *sortTransformation2) processView(mh *sortTableMergeHeap, cr flux.ColReader) error {
+	if cr.Len() == 0 {
+		return nil
+	}
+
+	cr.Retain()
+	item := &sortTableMergeHeapItem{cr: cr}
+	if !s.isSorted(cr, mh.sortCols) {
+		item.indices = s.sort(cr, mh.sortCols)
+		item.offset = int(item.indices.Value(0))
+	}
+	mh.items = append(mh.items, item)
```

Comment:
Wondering if it's possible here to check if the first value in `item` is in sorted order when compared against the last value in the last item of `mh.items`. Then we would know that across the entire table we are in sorted order, and the merge can simply be a concatenation of the buffers.

Reply:
I might not understand this completely so I'm going to restate what I think you're saying just in case I misunderstood.

In this section of code, we are merging buffers that have already been sorted. You are wondering if we can speed this up further if we know that all of the values in buffer 0 are less than buffer 1 and then we can append or use the buffer from buffer 0 and then buffer 1.

I think this is a good thing to try, but I suspect we won't get much of a performance improvement from it. The sort buffers don't necessarily have constructed buffers. They have a set of indices for how to sort the array. If the original buffer is sorted, then we have a sorted buffers. This might happen when we have `from |> range |> group`. But in this case, the buffers likely need to be merged so I don't think this matches the scenario you mentioned.

If the original buffers are unsorted, we have to reconstruct the arrays so using the original buffer doesn't help. If the original buffers are sorted, we could potentially benefit by using the original arrays. I think this optimization in this way would require the buffers to be sorted internally and then those buffers to be sorted so something like the output of `from |> range |> filter` would qualify but also it doesn't seem like you would use a sort at that moment. It also seems like we could make the planner smarter to remove the sort in these cases.

Another thing I've noticed is that sometimes copying data ends up being faster for future transformations even if it slows down the current one. The group transformation has this problem. It sometimes makes the buffers too small and it causes later transformations to hit worst case algorithm scenarios and prevents them from benefiting from cache locality. In these cases, it's sometimes better to reconstruct the buffer. But, it's also important that the buffer isn't too large to avoid too large of allocations. We'll have to experiment to find a balance.

I did wonder if doing this would potentially make reconstructing the arrays faster but I suspect that in the scenario where one buffer is completely less than another buffer, `heap.Fix(0)` will be just as fast because it would be a `O(1)` algorithmically. We might end up in a scenario though where instead of performing a comparison for each row we might be able to only perform one comparison. If this is what you mean, then I definitely think we should try that. I don't know a specific query pattern that would construct this data pattern so it might not be worth checking for.